### ISSUE TITLE
Tokenizer: what to do with these  very hard examples?

### ISSUE THREAD

**Issue by nicolabertoldi** [ModernMT/MMT#96]
Very hard examples from TagBenchmark-v0.2

```
Input: You must remove the file <userID>_<workstationName>.xml from the folder
Input (after tokenization and tag-stripping: You must remove the file _ .xml from the folder
Reference: E necessario rimuovere il file <userID>_<workstationName>.xml dalla folder
Output (after de-tokenization and tag-re-insertion): E necessario rimuovere il file <userID>_ <workstationName>.xml dalla folder
```

```
Input: <MSISDN>_<DATA>_<FLAGWOS>_<CCWOS>_<VLR>.frr
Input (after tokenization and tag-stripping: _ _ _ _ .frr
Reference: <MSISDN>_<DATA>_<FLAGWOS>_<CCWOS>_<VLR>.frr
Output (after de-tokenization and tag-re-insertion): <MSISDN>_ <DATA>_ <FLAGWOS>_ <CCWOS>_ <VLR>.frr
```

```
Input: WinNT://DOMAINNAME/USERNAME,user</log>
Input (after tokenization and tag-stripping:  WinNT : / / DOMAINNAME / USERNAME , user
Reference: WinNT://DOMAINNAME/USERNAME,user</log>
Output (after de-tokenization and tag-re-insertion): WinNT: / / DOMAINNAME / USERNAME, user</log>
```

**Comment by marcotrombetti** [comment]
@nicolabertoldi Why not simply removing double spaces?

**Comment by davidecaroselli** [comment]
To correctly handle this issue we need to introduce a feature in the post-processor that transfers from source sentence to target translation the "special" spacing decisions the input text has.

The current pipeline is ready to develop such feature thanks to the work done by @luca-mastrostefano (reversable pipeline). Thus I think this features has, for now, low priority so I move this issue for a future release.

**Comment by davidecaroselli** [comment]
@nicolabertoldi we have finally solved this issue :)

With commit https://github.com/ModernMT/MMT/commit/b194a8d72e612d5ada4df48ba0010b172b3cdffd we have introduced the `WhitespaceProjector`. What this component does is basically report the whitespaces from the source sentence to the target translation, with an heuristic that guess when the user has intentionally put a whitespace (or omitted one).

For example if the source sentence is "`this,that`", this component is able to understand that after the comma the user has intentionally removed a whitespace, so the translation will be outputted like so: "`questo,quello`".
Similarly if the user has forced a whitespace like in the sentence "`Here's a colon :`", the translation will be "`Ecco i due punti :`" and not "`Ecco i due punti:`".

The new version of MMT trained with benchmark 1.1  tested on those examples returned the correct de-tokenized text and gave us a nice boost of 0.1 BLEU points.
